src/toolkits/activity_classification/ac_data_iterator.cpp

```diff
@@ -132,152 +132,291 @@ variant_map_type _activity_classifier_prepare_data_impl(const gl_sframe &data,
+  // converted batch-processing ready format.
+  std::vector<std::string> output_column_names = {"features", "chunk_len",
+                                                  "session_id"};
+  std::vector<flex_type_enum> output_column_types = {flex_type_enum::VECTOR,
+                                                     flex_type_enum::INTEGER,
+                                                     data[session_id].dtype()};
+  if (use_target) {
+    output_column_names.insert(output_column_names.end(), {"target"});
+    output_column_types.insert(output_column_types.end(),
+                               {flex_type_enum::VECTOR});
+  }
 
-        time_t now = time(0);
-        if (verbose && difftime(now, last_print_time) > 10) {
-            logprogress_stream << "Pre-processing: " << std::setw(3) << (100 * processed_lines / data.size()) << "% complete"  << std::endl;
-            last_print_time = now;
+  gl_sframe_writer output_writer(output_column_names, output_column_types, 1);
+
+  if (verbose) {
+    logprogress_stream << "Using sequences of size " << chunk_length
+                       << " for model creation." << std::endl;
+  }
+
+  time_t last_print_time = time(0);
+  size_t processed_lines = 0;
+
+  // Iterate over the user data. The features and targets are aggregated, and
+  // handled whenever a the ending of a session is reached.
+  size_t chunk_size = 0;
+  for (const auto &line : data.range_iterator()) {
+    auto curr_session_id = line[column_index_map[session_id]];
+
+    if (curr_session_id != last_session_id) {
+
+      // Write the aggregated data of the current chunk (which includes all
+      // the samples in the current session) as a single new vector in the
+      // converted SFrame, and init all aggregation vectors to begin a new
+      // chunk.
+      if (curr_chunk_features.size() > 0) {
+        if (use_target) {
+          output_writer.write({curr_chunk_features, chunk_size, last_session_id,
+                               curr_chunk_targets},
+                              0);
+        } else {
+          output_writer.write(
+              {curr_chunk_features, chunk_size, last_session_id}, 0);
         }
+        curr_chunk_features.clear();
+        curr_chunk_targets.clear();
+      }
 
-        processed_lines += 1;
+      chunk_size = 0;
+      last_session_id = curr_session_id;
+      number_of_sessions++;
     }
 
-    // Handle the tail of the data - the last few lines of the last chunk, which needs to be finalized.
-    if (curr_chunk_features.size() > 0) {
-        finalize_chunk(curr_chunk_features,
-                       curr_chunk_targets,
-                       curr_window_targets,
-                       last_session_id,
-                       output_writer,
-                       chunk_size,
-                       feature_size,
-                       predictions_in_chunk,
-                       use_target);
+    chunk_size += 1;
+
+    for (const auto feature_name : features) {
```

Comment:
Use `const std::string&` instead of `const auto`

Using a reference prevents a copy in some cases. `std::string` is not much longer than `auto` and provides more clarity